dev/AppLifecycle/Association.cpp

```diff
@@ -0,0 +1,423 @@
+            0, samDesired, registeredAppsKey.put()));
+
+        DWORD pathSize{ 0 };
+        THROW_IF_WIN32_ERROR(::RegQueryValueEx(registeredAppsKey.get(), appId.c_str(), 0, nullptr,
+            nullptr, &pathSize));
+
+        std::wstring applicationKeyPath(pathSize, L'\0');
+        THROW_IF_WIN32_ERROR(::RegQueryValueEx(registeredAppsKey.get(), appId.c_str(), 0, nullptr,
+            reinterpret_cast<BYTE*>(applicationKeyPath.data()),
+            &pathSize));
+
+        wil::unique_hkey applicationKey;
+        THROW_IF_WIN32_ERROR(::RegOpenKeyEx(GetRegistrationRoot(), applicationKeyPath.c_str(),
+            0, samDesired, applicationKey.put()));
+
+        return applicationKey;
+    }
+
+    void RegisterApplication(const std::wstring& appId)
+    {
+        CreateApplicationKey(appId);
+    }
+
+    void UnregisterApplication(const std::wstring& appId)
+    {
+        std::wstring capabilitiesKeyPath = c_applicationsKeyPath + appId + c_capabilitiesKeyPath;
+        ::RegDeleteTree(GetRegistrationRoot(), capabilitiesKeyPath.c_str());
+
+        wil::unique_hkey registeredAppsKey;
+        if (::RegOpenKeyEx(GetRegistrationRoot(), c_registeredApplicationsKeyPath.c_str(), 0,
+            KEY_WRITE, registeredAppsKey.put()) == ERROR_SUCCESS)
+        {
+            ::RegDeleteValue(registeredAppsKey.get(), appId.c_str());
+        }
+    }
+
+    void RegisterVerb(const std::wstring& progId, const std::wstring& verb, const std::wstring& command,
+        _In_opt_ const GUID* delegateExecute)
+    {
+        auto assocKey = OpenAssocKey(progId, KEY_WRITE);
+        auto key_path = c_shellKeyName + L"\\" + verb + L"\\" + c_commandKeyName;
+        wil::unique_hkey key;
+
+        THROW_IF_WIN32_ERROR(::RegCreateKeyEx(assocKey.get(), key_path.c_str(), 0, nullptr, 0,
+            KEY_WRITE, nullptr, key.put(), nullptr));
+
+        THROW_IF_WIN32_ERROR(::RegSetValueEx(key.get(), nullptr, 0, REG_SZ,
+            reinterpret_cast<BYTE const*>(command.c_str()),
+            static_cast<uint32_t>((command.size() + 1) * sizeof(wchar_t))));
+
+        if (delegateExecute)
+        {
+            std::wstring delegateClsid{ LR"({????????-????-????-????-????????????})" };
+            ::StringFromGUID2(*delegateExecute, delegateClsid.data(), 39);
+
+            THROW_IF_WIN32_ERROR(::RegSetValueEx(key.get(), c_delegateExecuteValueName.c_str(), 0, REG_SZ,
+                reinterpret_cast<BYTE const*>(delegateClsid.c_str()),
+                static_cast<uint32_t>((delegateClsid.size() + 1) * sizeof(wchar_t))));
+        }
+    }
+
+    void UnregisterVerb(const std::wstring& progId, const std::wstring& verb)
+    {
+        auto key_path = CreateAssocKeyPath(progId) + L"\\" + c_shellKeyName + L"\\" + verb +
+            L"\\" + c_commandKeyName;
+        ::RegDeleteTree(GetRegistrationRoot(), key_path.c_str());
+    }
+
+    void RegisterProtocol(const std::wstring& scheme)
```

Comment:
in many of the functions in this section, I find myself wanting a short comment that shows an example of the string that is being built.